## Data-analysis
Banco de dados referente a quantidade de empresas abertas na Cidade de Langley/BC no periodo de Dez 2021 a Marco 2022. 

##IMPORTACAO DOS DADOS

  Importar a tabela .csv

##TRATAMENTO DOS DADOS - obs.: muito dos recursos utlizados aqui nao seriam necessarios para responder as questoes, mas foram feitos no sentido de praticar e conhecer mais recursos.

    Organizar a planilha 
    Usar varios recuros para entender a tabela
    Deletar linhas e colunas nao necessarias
    Identificar celulas vazias
    Etc
    
##ANALISE DOS DADOS
   
1. Quantas empresas foram abertas ao todo? 
= 7019

2. Quantas empresas foram abertas por area/local?
Willoughby                982
Willowbrook               634
Walnut Grove              492
Aldergrove                490
Brookswood / Fernridge    412
Murrayville               406
Northwest Langley         289
Gloucester                256
Fort Langley              248

3. Quais sao os top 3 por tipo de empresas que mais pediram abertura de empresa?
= Construcao (residential & general) seguida de filme dividos assim:
Tipo                                            Qtde empresas abertas
RESIDENTIAL GENERAL CONTRACTOR                   120
MOTION PICTURE PRODUCTION                        110
GENERAL CONTRACTOR                                86

4. Quantos funcionarios estao empregados (ao todo)?
= 75963

5. Quantos funcionarios estao trabalhando por tipo de empregabilidade?
Seazonal = 4961, Part-time = 12791 and Full-Time = 58211. A maioria dos empregados na cidade de Langley trabalham full-time.

6. Ano/periodo que essas empresas foram abertas?


In [1]:
#1. Importar a tabela .csv

import pandas as pd
import numpy as np

df = pd.read_csv("Business_Licenses.csv")
display(df)


X            Y   FID  PropertyNumber  \
0    -13655110.99  6302015.650     1          150466   
1             NaN          NaN     2          136578   
2    -13654577.37  6305261.856     3          149216   
3    -13650665.24  6291711.370     4          102680   
4    -13650331.94  6290096.797     5          109491   
...           ...          ...   ...             ...   
7014 -13655593.26  6287954.262  7015          120750   
7015 -13633082.68  6284648.347  7016          122124   
7016 -13640632.83  6286425.635  7017          133634   
7017 -13633429.94  6284976.631  7018          133290   
7018 -13650508.65  6296054.182  7019          110595   

                                Business_Name Civic_Unit  Civic_House  \
0      LINDSAY KENNEY BARRISTERS & SOLICITORS        400       8621.0   
1                   SITKA FIRE PROTECTION INC        NaN          NaN   
2                   WALNUT GROVE BOTTLE DEPOT          6      20280.0   
3          CANADIAN MUSEUM FLIGHT ASSOCIATION          3       5333.0   
4         MURRAYVILLE COMMUNITY MEMORIAL HALL        NaN      21667.0   
...                                       ...        ...          ...   
7014                   CEDARBROOKS BAKERY LTD        NaN       4123.0   
7015  STEVE'S CUSTOM HOME LIGHTING (1982) LTD        NaN      27300.0   
7016           OTTER FARM & HOME CO-OPERATIVE        NaN       3650.0   
7017               PEARCE HARDWARE (1977) LTD        NaN       3147.0   
7018                         NAPIER SALES LTD        NaN       6743.0   

     Civic_Street                                  Mailing_Address   GISLink  \
0          201 ST            400-8621 201 ST \nLANGLEY BC  V2Y 0G9  148253.0   
1     OUT OF TOWN      207 8475 ONTARIO ST \nVANCOUVER BC  V5X 3E8  136578.0   
2          97 AVE             8 20280 97 AVE \nLANGLEY BC  V1M 4B9  149216.0   
3          216 ST              3-5333 216 ST \nLANGLEY BC  V2Y 2N3  102680.0   
4          48 AVE               21667 48 AVE \nLANGLEY BC  V3A 3M7  109491.0   
...           ...                                              ...       ...   
7014       200 ST                4123 200 ST \nLANGLEY BC  V3A 1K8  114407.0   
7015   FRASER HWY        27294 FRASER HWY \nALDERGROVE BC  V4W 3P8  112734.0   
7016       248 ST                3650 248 ST \nLANGLEY BC  V4W 1X7  133634.0   
7017       272 ST  BOX 1119 \n3147 272 ST \nALDERGROVE BC  V4W 2V1  133290.0   
7018       216 ST                6743 216 ST \nLANGLEY BC  V2Y 0W9  110595.0   

      ... NAICS_Primary                                 NAICS_Primary_Desc  \
0     ...      541110.0  Professional, scientific & technical services ...   
1     ...      238220.0  Construction / Specialty trade contractors / B...   
2     ...      418190.0  Wholesale trade / Miscellaneous wholesaler-dis...   
3     ...      712119.0  Arts, entertainment & recreation / Heritage in...   
4     ...      813410.0  Other services (exc. Public administration) / ...   
...   ...           ...                                                ...   
7014  ...      311811.0  Manufacturing / Food mfg. / Bakeries & tortill...   
7015  ...      442110.0  Retail trade / Furniture & home furnishings st...   
7016  ...      452910.0  Retail trade / General merchandise stores / Ot...   
7017  ...      444130.0  Retail trade / Building material & garden equi...   
7018  ...      453310.0  Retail trade / Misc. store retailers / Used me...   

     Approved_Date Permitted_Use Permitted_Use_Desc Seasonal_Employees  \
0              NaN           261                NaN                NaN   
1              NaN           NaN                NaN                NaN   
2              NaN           289                NaN                NaN   
3              NaN           NaN                NaN                NaN   
4              NaN           NaN                NaN                NaN   
...            ...           ...                ...                ...   
7014           NaN            83                Na

In [2]:
df.columns   # lista todas as colunas
list(df.columns)

['X',
 'Y',
 'FID',
 'PropertyNumber',
 'Business_Name',
 'Civic_Unit',
 'Civic_House',
 'Civic_Street',
 'Mailing_Address',
 'GISLink',
 'CommunityName',
 'Phone',
 'Licence_Desc',
 'Category',
 'Category_Desc',
 'Detail',
 'Detail_Desc',
 'NAICS_Primary',
 'NAICS_Primary_Desc',
 'Approved_Date',
 'Permitted_Use',
 'Permitted_Use_Desc',
 'Seasonal_Employees',
 'PartTime_Employees',
 'FullTime_Employees',
 'Latitude',
 'Longitude']

##### A tabela tem 7019 linhas e 27 colunas. Analisando a lista de colunas nesta tabela, varias colunas podem ser deletadas
##### O proximo passo deleta-las para reduzir a tabela. 

In [3]:
# Deletar as colunas q nao necessarias

df = df.drop(["X", "Y", "FID", "PropertyNumber","Civic_Unit", "Civic_House", "Civic_Street", "Mailing_Address","Detail", "GISLink", "Phone", "NAICS_Primary","Category","Permitted_Use", "Permitted_Use_Desc", "Latitude", "Longitude"], axis = 1) 
display(df)

Business_Name           CommunityName  \
0      LINDSAY KENNEY BARRISTERS & SOLICITORS              Willoughby   
1                   SITKA FIRE PROTECTION INC                     NaN   
2                   WALNUT GROVE BOTTLE DEPOT       Northwest Langley   
3          CANADIAN MUSEUM FLIGHT ASSOCIATION             Murrayville   
4         MURRAYVILLE COMMUNITY MEMORIAL HALL             Murrayville   
...                                       ...                     ...   
7014                   CEDARBROOKS BAKERY LTD  Brookswood / Fernridge   
7015  STEVE'S CUSTOM HOME LIGHTING (1982) LTD              Aldergrove   
7016           OTTER FARM & HOME CO-OPERATIVE                     NaN   
7017               PEARCE HARDWARE (1977) LTD              Aldergrove   
7018                         NAPIER SALES LTD                     NaN   

                             Licence_Desc      Category_Desc  \
0                                LAW FIRM      PROFESSIONALS   
1     FIRE ALARM & WATER WORKS CONTRACTOR  CONTRACTOR - MISC   
2                RECYCLING & BOTTLE DEPOT  GARBAGE COLLECTOR   
3                           FLIGHT MUSEUM                NaN   
4                          COMMUNITY HALL         NON-PROFIT   
...                                   ...                ...   
7014                               RETAIL   RETAIL MERCHANTS   
7015                               RETAIL   RETAIL MERCHANTS   
7016    RETAIL FOOD HARDWARE AND DRYGOODS   RETAIL MERCHANTS   
7017                      RETAIL HARDWARE   RETAIL MERCHANTS   
7018          CHICK SALES/RETAIL ANTIQUES   RETAIL MERCHANTS   

              Detail_Desc                                 NAICS_Primary_Desc  \
0                  LAWYER  Professional, scientific & technical services ...   
1         FIRE PROTECTION  Construction / Specialty trade contractors / B...   
2     RECYCLING EQUIPMENT  Wholesale trade / Miscellaneous wholesaler-dis...   
3                     NaN  Arts, entertainment & recreation / Heritage in...   
4                 SOCIETY  Other services (exc. Public administration) / ...   
...                   ...                                                ...   
7014               BAKERY  Manufacturing / Food mfg. / Bakeries & tortill...   
7015       HARDWARE STORE  Retail trade / Furniture & home furnishings st...   
7016       HARDWARE STORE  Retail trade / General merchandise stores / Ot...   
7017       HARDWARE STORE  Retail trade / Building material & garden equi...   
7018    SECOND HAND STORE  Retail trade / Misc. store retailers / Used me...   

     Approved_Date  Seasonal_Employees  PartTime_Employees  FullTime_Employees  
0              NaN                 NaN                 5.0                30.0  
1              NaN                 NaN                 NaN                 3.0  
2              NaN                 NaN                 1.0                 5.0  
3              NaN                 NaN                 NaN                 NaN  
4              NaN                 NaN                 NaN                 NaN  
...            ...                 ...                 ...                 ...  
7014           NaN                 NaN                 NaN                 NaN  
7015           NaN                 NaN                 NaN                 NaN  
7016           NaN                 NaN                 NaN                 NaN  
7017           NaN                 1.0                 2.0                 4.0  
7018           NaN                 NaN                 3.0                 NaN  

[7019 rows x 10 columns]

##### Apos a deletion, restaram 9 colunas na tabela.  

In [4]:
df.dtypes  # aqui eh possivel ver como o Python categoriza o tipo de variavel. Approved_Date precisara ser trabalhado. 

Business_Name          object
CommunityName          object
Licence_Desc           object
Category_Desc          object
Detail_Desc            object
NAICS_Primary_Desc     object
Approved_Date          object
Seasonal_Employees    float64
PartTime_Employees    float64
FullTime_Employees    float64
dtype: object

In [5]:
# info          # forma de verificar celulas vazias - imprimira um resumo de qtos valores nao nulos estao na tabela
                # essa forma tem que contar o complementar para saber qtas estao vazias. Ai entra a outra forma df.isna

print(df.info()) 


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7019 entries, 0 to 7018
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Business_Name       7019 non-null   object 
 1   CommunityName       4209 non-null   object 
 2   Licence_Desc        7019 non-null   object 
 3   Category_Desc       7011 non-null   object 
 4   Detail_Desc         6914 non-null   object 
 5   NAICS_Primary_Desc  6853 non-null   object 
 6   Approved_Date       255 non-null    object 
 7   Seasonal_Employees  4463 non-null   float64
 8   PartTime_Employees  5130 non-null   float64
 9   FullTime_Employees  6286 non-null   float64
dtypes: float64(3), object(7)
memory usage: 548.5+ KB
None


##### A tabela acima (info) mostra a qty de celulas vazias para cada coluna. 
#####Ex.: P/coluna "CommunityName" -> total entries(= 7019) menos 4209 non-null = 2810 nao nulos
#####Ex.: P/coluna "Licence_Desc" -> total entries(= 7019) menos 7019 non-null = zero celulas vazias

##### Essa "info" tem a vantagem de descrever o tipo de variavel: 
##### Ex que chama a atencao "Approved_Date" esta sendo reconhecido como object, mas deveria ser data

##### Alterando o tipo de variavel Object to data --> Approved_Date 
##### Se fosse numerico seria: df["nome da coluna"] = pd.to_numeric(df["nome da coluna"], erros="coerce")
##### coerce eh um codigo que na hora que houver a transformacao do tipo da variavel, o coerce fara c/ q o campo fique vazio. 
##### df["Approved_Date"] = pd.to_datetime(df["Approved_Date"], erros="coerce")

In [6]:
# df.isna().any()    or .sum()               # outra forma de verificar celulas vazias "isna -> eh vazio"

df.isna().any()
print(df.isna())


      Business_Name  CommunityName  Licence_Desc  Category_Desc  Detail_Desc  \
0             False          False         False          False        False   
1             False           True         False          False        False   
2             False          False         False          False        False   
3             False          False         False           True         True   
4             False          False         False          False        False   
...             ...            ...           ...            ...          ...   
7014          False          False         False          False        False   
7015          False          False         False          False        False   
7016          False           True         False          False        False   
7017          False          False         False          False        False   
7018          False           True         False          False        False   

      NAICS_Primary_Desc  Approved_Date

In [7]:
# df.isna().any()    or .sum()         # Outra forma de verificar celulas vazias "isna -> is empty"
                                       # essa ja ajuda mais e mostra que ha muitas celulas sem preenchimento.
                                       # Tentaria descobrir a razao p/ encontrar uma melhoria no processo de coleta.
#df.isna().sum()
print(df.isna().sum())


Business_Name            0
CommunityName         2810
Licence_Desc             0
Category_Desc            8
Detail_Desc            105
NAICS_Primary_Desc     166
Approved_Date         6764
Seasonal_Employees    2556
PartTime_Employees    1889
FullTime_Employees     733
dtype: int64


##### A forma acima eh mais direta p contar qtas celulas estao vazias. Ou seja, tem mtas celulas vazias, o que chama atencao.

In [8]:
#Identificar alguns possiveis preenchimentos c vazios n hora de importar. Nesse caso nda mudara pq n ha essa situacao
#Valores_vazios = ["?", "-"]
#df = pd.read_csv("Business_Licenses.csv", na_values=valores_vazios)

#identificar com replace
df = df.replace({"?": np.nan, "-": np.nan})
display(df)


Business_Name           CommunityName  \
0      LINDSAY KENNEY BARRISTERS & SOLICITORS              Willoughby   
1                   SITKA FIRE PROTECTION INC                     NaN   
2                   WALNUT GROVE BOTTLE DEPOT       Northwest Langley   
3          CANADIAN MUSEUM FLIGHT ASSOCIATION             Murrayville   
4         MURRAYVILLE COMMUNITY MEMORIAL HALL             Murrayville   
...                                       ...                     ...   
7014                   CEDARBROOKS BAKERY LTD  Brookswood / Fernridge   
7015  STEVE'S CUSTOM HOME LIGHTING (1982) LTD              Aldergrove   
7016           OTTER FARM & HOME CO-OPERATIVE                     NaN   
7017               PEARCE HARDWARE (1977) LTD              Aldergrove   
7018                         NAPIER SALES LTD                     NaN   

                             Licence_Desc      Category_Desc  \
0                                LAW FIRM      PROFESSIONALS   
1     FIRE ALARM & WATER WORKS CONTRACTOR  CONTRACTOR - MISC   
2                RECYCLING & BOTTLE DEPOT  GARBAGE COLLECTOR   
3                           FLIGHT MUSEUM                NaN   
4                          COMMUNITY HALL         NON-PROFIT   
...                                   ...                ...   
7014                               RETAIL   RETAIL MERCHANTS   
7015                               RETAIL   RETAIL MERCHANTS   
7016    RETAIL FOOD HARDWARE AND DRYGOODS   RETAIL MERCHANTS   
7017                      RETAIL HARDWARE   RETAIL MERCHANTS   
7018          CHICK SALES/RETAIL ANTIQUES   RETAIL MERCHANTS   

              Detail_Desc                                 NAICS_Primary_Desc  \
0                  LAWYER  Professional, scientific & technical services ...   
1         FIRE PROTECTION  Construction / Specialty trade contractors / B...   
2     RECYCLING EQUIPMENT  Wholesale trade / Miscellaneous wholesaler-dis...   
3                     NaN  Arts, entertainment & recreation / Heritage in...   
4                 SOCIETY  Other services (exc. Public administration) / ...   
...                   ...                                                ...   
7014               BAKERY  Manufacturing / Food mfg. / Bakeries & tortill...   
7015       HARDWARE STORE  Retail trade / Furniture & home furnishings st...   
7016       HARDWARE STORE  Retail trade / General merchandise stores / Ot...   
7017       HARDWARE STORE  Retail trade / Building material & garden equi...   
7018    SECOND HAND STORE  Retail trade / Misc. store retailers / Used me...   

     Approved_Date  Seasonal_Employees  PartTime_Employees  FullTime_Employees  
0              NaN                 NaN                 5.0                30.0  
1              NaN                 NaN                 NaN                 3.0  
2              NaN                 NaN                 1.0                 5.0  
3              NaN                 NaN                 NaN                 NaN  
4              NaN                 NaN                 NaN                 NaN  
...            ...                 ...                 ...                 ...  
7014           NaN                 NaN                 NaN                 NaN  
7015           NaN                 NaN                 NaN                 NaN  
7016           NaN                 NaN                 NaN                 NaN  
7017           NaN                 1.0                 2.0                 4.0  
7018           NaN                 NaN                 3.0                 NaN  

[7019 rows x 10 columns]

#####Removendo valores vazios ( axis = 0 -> exclui e axis = 1 -> exclui coluna/ 
##### how= "any" -> exclui pelo menos uma celula que tem NaN   e "all" todas as celulas da linha tem que ser NaN
#####df = df.dropna(how="any", axis=0)
#####display(df)
#####Como na tabela ha muitos valores NaN, essa exclusao nao fara sentido. 

In [9]:
freq_type_bus1 = df["Detail_Desc"].value_counts()
display(freq_type_bus1)


NOT USED                    493
SINGLE FAMILY DWELLING      292
MISCELLANEOUS               175
LANDSCAPING                 172
RESIDENTIAL & COMMERCIAL    153
                           ... 
TEMP.CONTROLS                 1
VACUUM INSTALLATION           1
CHIMNEY                       1
BOWLING ALLEY                 1
PSYCHIATRIST                  1
Name: Detail_Desc, Length: 480, dtype: int64

## ANALISE DOS DADOS 

In [10]:
# 1.  Quantas empresas foram abertas ao todo?

freq_bus_name = df["Business_Name"].count()
display(freq_bus_name)  

7019

In [11]:
# 2.  Quantas empresas foram abertas por area/local?

freq_bus_area = df["CommunityName"].value_counts()
display(freq_bus_area)


Willoughby                982
Willowbrook               634
Walnut Grove              492
Aldergrove                490
Brookswood / Fernridge    412
Murrayville               406
Northwest Langley         289
Gloucester                256
Fort Langley              248
Name: CommunityName, dtype: int64

In [12]:
# 3. Quais sao os top 3 por tipo de empresas que mais pediram abertura de empresa?
freq_type_bus = df["Licence_Desc"].value_counts()
display(freq_type_bus)
                                                            # Construcao (residential & general) seguida de filme

RESIDENTIAL GENERAL CONTRACTOR          120
MOTION PICTURE PRODUCTION               110
GENERAL CONTRACTOR                       86
ELECTRICAL CONTRACTOR                    79
REGISTERED MASSAGE THERAPIST             76
                                       ... 
BROKER FOR WHOLESALE GARDEN SUPPLIES      1
WHOLESALE OF BUILDING PRODUCTS            1
ENVIRONMENTAL ENGINEERING CONSULTANT      1
OFFICE FOR CONSTRUCTION OF NEW HOMES      1
INTERIOR RENOVATION CONTRACTOR            1
Name: Licence_Desc, Length: 4939, dtype: int64

In [13]:
# 4. Quantos funcionarios estao empregados (ao todo)?

total = round((df["Seasonal_Employees"].sum() + df["PartTime_Employees"].sum() + df["FullTime_Employees"].sum()))
print(total)


75963


In [14]:
# 5. Quantos funcionarios estao trabalhando por tipo de empregabilidade?
freq_seas_emp = round(df["Seasonal_Employees"].sum())
display(freq_seas_emp)

freq_part_emp = round(df["PartTime_Employees"].sum())
display(freq_part_emp)

freq_full_emp = round(df["FullTime_Employees"].sum())
display(freq_full_emp)



4961

12791

58211

In [15]:
# 6.  Ano/periodo que essas empresas foram abertas?
df["Approved_Date"].unique()  


array([nan, '2022/03/15 00:00:00+00', '2022/01/25 00:00:00+00',
       '2022/01/17 00:00:00+00', '2022/01/27 00:00:00+00',
       '2022/01/26 00:00:00+00', '2022/02/09 00:00:00+00',
       '2022/02/18 00:00:00+00', '2022/01/24 00:00:00+00',
       '2022/01/04 00:00:00+00', '2022/03/01 00:00:00+00',
       '2022/01/11 00:00:00+00', '2022/01/12 00:00:00+00',
       '2022/01/20 00:00:00+00', '2022/01/18 00:00:00+00',
       '2022/01/14 00:00:00+00', '2022/03/14 00:00:00+00',
       '2022/01/10 00:00:00+00', '2022/01/05 00:00:00+00',
       '2021/12/24 00:00:00+00', '2022/01/28 00:00:00+00',
       '2022/01/21 00:00:00+00', '2022/02/07 00:00:00+00',
       '2021/12/23 00:00:00+00', '2022/01/06 00:00:00+00',
       '2022/03/02 00:00:00+00', '2022/01/13 00:00:00+00',
       '2022/03/10 00:00:00+00', '2022/02/11 00:00:00+00',
       '2022/01/19 00:00:00+00', '2022/02/25 00:00:00+00',
       '2022/02/02 00:00:00+00', '2022/02/08 00:00:00+00',
       '2022/03/11 00:00:00+00', '2022/02/04 00:00:

In [36]:
df1 = df.sort_values(by="Approved_Date").dropna()
df1["Approved_Date"].unique()[0]


'2021/12/23 00:00:00+00'

In [37]:
df1["Approved_Date"].unique()[1]


'2021/12/24 00:00:00+00'

In [39]:
df1["Approved_Date"].unique()[-1]

'2022/03/18 00:00:00+00'